langchain/callbacks/base.py

```diff
@@ -252,3 +258,287 @@ def remove_handler(self, handler: BaseCallbackHandler) -> None:
         self.handlers = handlers
+
+
+class AsyncCallbackHandler(BaseCallbackHandler):
+    """Async callback handler that can be used to handle callbacks from langchain."""
+
+    async def on_llm_start(
+        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
+    ) -> None:
+        """Run when LLM starts running."""
+
+    async def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
+        """Run on new LLM token. Only available when streaming is enabled."""
+
+    async def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
+        """Run when LLM ends running."""
+
+    async def on_llm_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> None:
+        """Run when LLM errors."""
+
+    async def on_chain_start(
+        self, serialized: Dict[str, Any], inputs: Dict[str, Any], **kwargs: Any
+    ) -> None:
+        """Run when chain starts running."""
+
+    async def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:
+        """Run when chain ends running."""
+
+    async def on_chain_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> None:
+        """Run when chain errors."""
+
+    async def on_tool_start(
+        self, serialized: Dict[str, Any], action: AgentAction, **kwargs: Any
+    ) -> None:
+        """Run when tool starts running."""
+
+    async def on_tool_end(self, output: str, **kwargs: Any) -> None:
+        """Run when tool ends running."""
+
+    async def on_tool_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> None:
+        """Run when tool errors."""
+
+    async def on_text(self, text: str, **kwargs: Any) -> None:
+        """Run on arbitrary text."""
+
+    async def on_agent_finish(self, finish: AgentFinish, **kwargs: Any) -> None:
+        """Run on agent end."""
+
+
+class AsyncCallbackManager(BaseCallbackManager):
+    """Async callback manager that can be used to handle callbacks from LangChain."""
+
+    @property
+    def is_async(self) -> bool:
+        """Return whether the handler is async."""
+        return True
+
+    def __init__(self, handlers: List[BaseCallbackHandler]) -> None:
+        """Initialize callback manager."""
+        self.handlers: List[BaseCallbackHandler] = handlers
+
+    async def on_llm_start(
+        self,
+        serialized: Dict[str, Any],
+        prompts: List[str],
+        verbose: bool = False,
+        **kwargs: Any
+    ) -> None:
+        """Run when LLM starts running."""
+        for handler in self.handlers:
+            if not handler.ignore_llm:
+                if verbose or handler.always_verbose:
+                    if asyncio.iscoroutinefunction(handler.on_llm_start):
+                        await handler.on_llm_start(serialized, prompts, **kwargs)
+                    else:
+                        await asyncio.get_event_loop().run_in_executor(
```

Comment:
@nfcampos -> Likewise, I'm not sure this is the best approach, as the handler is not guaranteed to be thread safe

Reply:
An idea, what if AsyncCallbackManager holds an instance of a sync SharedCallbackManager, and when a sync handler is added, it gets added to that one instead (transparently to the user). Then in the on_* methods, it goes something like

```
async def on_llm_start(...):
    for handler in self.handlers:
         if not handler.ignore_llm:
             if verbose or handler.always_verbose:
                  await handler.on_llm_start(serialized, prompts, **kwargs)
    await asyncio.get_event_loop().run_in_executor(None, self.sync_shared_callback_manager. on_llm_start, ...)
```

And it is thread safe because SharedCallbackManager makes it so?